org.uqbar.project.wollok.lib/src/wollok/lang/WNumber.xtend

```diff
@@ -1,63 +1,165 @@
+	def dispatch Number doPlus(BigDecimal w) {
+		this.add(wrapped, w)
+	}
 
-	def min(WNumber<?> other) { doMin(this, other).asWollokObject }
+	// TODO: here it should throw a 100% wollok exception class
+	def dispatch Number doPlus(Object w) { throw new WollokRuntimeException("Cannot add " + w) }
+
+	@NativeMessage("-")
+	def minus(WollokObject other) { operate(other)[doMinus(it)] }
+
+	def dispatch Number doMinus(BigDecimal w) {
+		this.subtract(wrapped, w)
+	}
+
+	def dispatch Number doMinus(Object w) { throw new WollokRuntimeException("Cannot substract " + w) }
+
+	@NativeMessage("*")
+	def multiply(WollokObject other) { operate(other)[doMultiply(it)] }
+
+	def dispatch Number doMultiply(BigDecimal w) {
+		this.mul(wrapped, w)
+	}
 
-	def dispatch doMin(WInteger a, WInteger b) { Math.min(a.wrapped, b.wrapped) }
+	def dispatch Number doMultiply(Object w) { throw new WollokRuntimeException("Cannot multiply " + w) }
 
-	def dispatch doMin(WNumber<?> a, WNumber<?> b) { Math.min(a.doubleValue, b.doubleValue) }
+	@NativeMessage("/")
+	def divide(WollokObject other) { operate(other)[doDivide(it)] }
+
+	def dispatch Number doDivide(BigDecimal w) {
+		this.div(wrapped, w)
+	}
+
+	def dispatch Number doDivide(Object w) { throw new WollokRuntimeException("Cannot divide " + w) }
+
+	@NativeMessage("**")
+	def raise(WollokObject other) { operate(other)[doRaise(it)] }
+
+	def dispatch Number doRaise(BigDecimal w) { Math.pow(wrapped.doubleValue, w.doubleValue) }
+
+	def dispatch Number doRaise(Object w) { throw new WollokRuntimeException("Cannot raise " + w) }
+
+	@NativeMessage("%")
+	def module(WollokObject other) { operate(other)[doModule(it)] }
+
+	def dispatch Number doModule(BigDecimal w) {
+		this.remainder(wrapped, w)
+	}
+
+	def dispatch Number doModule(Object w) { throw new WollokRuntimeException("Cannot module " + w) }
+
+	def abs() { this.wrapped.abs.asWollokObject }
+
+	def invert() { (-wrapped).asWollokObject }
+
+	def randomUpTo(WollokObject other) {
+		val maximum = other.nativeNumber.wrapped.doubleValue
+		val minimum = wrapped.doubleValue()
+		((Math.random * (maximum - minimum)) + minimum).doubleValue()
+	}
+
+	def roundUp(int decimals) {
+		scale(decimals, BigDecimal.ROUND_UP)
+	}
+
+	def truncate(int decimals) {
+		scale(decimals, BigDecimal.ROUND_DOWN)
+	}
+
+	@NativeMessage(">")
+	def greater(WollokObject other) { wrapped.doubleValue > other.nativeNumber.wrapped.doubleValue }
+
+	@NativeMessage(">=")
+	def greaterOrEquals(WollokObject other) { wrapped.doubleValue >= other.nativeNumber.wrapped.doubleValue }
+
+	@NativeMessage("<")
+	def lesser(WollokObject other) { wrapped.doubleValue < other.nativeNumber.wrapped.doubleValue }
+
+	@NativeMessage("<=")
+	def lesserOrEquals(WollokObject other) { wrapped.doubleValue <= other.nativeNumber.wrapped.doubleValue }
+
+	def gcd(WollokObject other) {
+		if (!wrapped.isInteger) {
+			throw new IllegalArgumentException("gcd expects receiver to be an integer")
```

Comment:
I would prefer "gcd can only be applied to integer numbers".

Also this should be a pure Wollok exception, but that problem is already reported in this file, so I will create an issue to solve it later.